src/main/java/com/redhat/pantheon/jcr/JcrQueryHelper.java

```diff
@@ -0,0 +1,116 @@
+package com.redhat.pantheon.jcr;
+
+import com.google.common.collect.Lists;
+import org.apache.sling.api.resource.Resource;
+import org.apache.sling.api.resource.ResourceResolver;
+
+import javax.jcr.RepositoryException;
+import javax.jcr.Session;
+import javax.jcr.query.Query;
+import javax.jcr.query.QueryManager;
+import javax.jcr.query.QueryResult;
+import javax.jcr.query.Row;
+import javax.jcr.query.qom.Ordering;
+import javax.jcr.query.qom.QueryObjectModel;
+import javax.jcr.query.qom.QueryObjectModelFactory;
+import javax.jcr.query.qom.Selector;
+import java.util.Iterator;
+import java.util.stream.Stream;
+
+/**
+ * Provides utility methods for querying the Sling JCR repository.
+ */
+public class JcrQueryHelper {
+
+    private final ResourceResolver resourceResolver;
+
+    public JcrQueryHelper(ResourceResolver resourceResolver) {
+        this.resourceResolver = resourceResolver;
+    }
+
+    /**
+     * A convenience method to run a query against the JCR Repository adding limit and offset values useful for
+     * pagination.
+     * @param query A full JCR SQL2 query
+     * @param limit the number of results to return (for pagination)
+     * @param offset i.e. How many results to skip (for pagination)
+     * @return A stream of sling resources resulting from the query
+     * @throws RepositoryException
+     */
+    public Stream<Resource> query(String query, long limit, long offset)
+            throws RepositoryException {
+        Session session = resourceResolver.adaptTo(Session.class);
+
+        QueryManager queryManager = session.getWorkspace().getQueryManager();
+        Query queryObj = queryManager.createQuery(query, Query.JCR_SQL2);
+        queryObj.setLimit(limit);
+        queryObj.setOffset(offset);
```

Comment:
There are pitfalls with that approach. For example, we would need to start managing session state, which is a pain. My preferred approach at this point would be manage page state in the client (e.g. current page number) and then have the server fetch each page when requested.

Reply:
Yeah, makes perfect sense. Especially since we'll eventually be deploying this on OpenShift, and it would be near impossible to synchronize state between pods. The only payoff would be less hits to the database, but that's a good candidate for "don't solve problems we don't have yet." Consider my curiosity resolved.